File: table.py
import sys
from utils import R, seedGen, keyGen, alphabet, encriptV1, encriptV2

if len(sys.argv) != 4:
	print("Missing args! Please run the application with the following arguments:")
	print("python3 {} <password length> <rainbow table size> <output file name>".format(sys.argv[0]))
	exit()

#Args
rainbowFileName = sys.argv[3]
l = int(sys.argv[1])
s = int(sys.argv[2])

#Validations (some)
if l < 1 or l > 16:
	print("Password length not valid!")
	exit()
if s <1:
	print("Rainbow table size not valid!")
	exit()

#Variables
alphabetLen = len(alphabet)
numRows = int((16*(2**s))/(l*2)) 
k = (int((alphabetLen**l)/numRows)+1)*2

print("Output file: {}".format(rainbowFileName))
print("Num rows: {}".format(numRows))
print("Password length: {}".format(l))
print("s value: {}".format(s))
print("k value: {}".format(k))

#Generate RAINBOW TABLE
usedSeeds = []
lastPwds = []
for row in range(0,numRows):
	while True:
		seed = seedGen(alphabet, l)
		if seed not in usedSeeds:
			usedSeeds.append(seed)
			break

	pwd = seed
	for ite in range(0, k):
		key = bytes(keyGen(pwd), 'UTF-8')
		#pwd = encriptV1(key, l, ite)#Code provided in classe but not efficient
		pwd, h = encriptV2(key, l, ite)
	lastPwds.append(pwd)
	print("Interaction {} of {}".format(row, numRows))

#This for is not time consuming
file = open(rainbowFileName, 'wb')
file.write(bytes(str(l), 'UTF-8'))
file.write(bytes(str(k), 'UTF-8'))
file.write(bytes("\n", 'UTF-8'))
for seed, pwd in zip(usedSeeds, lastPwds):
	file.write(bytes(seed, 'UTF-8'))
	file.write(bytes(pwd, 'UTF-8'))
file.close()

File: table.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <math.h>
#include <tgmath.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/aes.h>
#include <openssl/rand.h>
#include <collectc/hashtable.h>
#include "hashtable.h"
#define  KEY_LEN  16 /*  bytes*/

const EVP_CIPHER *h;
EVP_CIPHER_CTX *ctx;
const char *alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789?!";//[a-zA-Z0-9?!]
const int alphabetLen = 64;
int l;
long rows;

void getPwd(char* pwd, long pwdIndex)
{
	long value = pwdIndex;
	for(int x = l-1; x >= 0; x--)
	{
		int charIndex = value % alphabetLen;
		value = (long)value / alphabetLen;
		pwd[x] = alphabet[charIndex];
	}
	pwd[l] = '\0';
}

void keyGen(char* key, char* pwd)
{
	int pwdCounter = 0;
 	for (int i = 0; i < 16; i++)
 	{
    	if (pwdCounter == l)
 			pwdCounter = 0;
        key[i] = pwd[pwdCounter];
    	pwdCounter++;
    }
    key[16] = '\0';
}


int main(int argc, char *argv[])
{
	/**
	 * Variables
	 */
	char *programName = argv[0];
	
	if (argc != 4)
	{
		printf("Missing args! Please run the application with the following arguments:\n");
		printf("%s <password length> <rainbow table size> <output file name>\n", programName);
		exit(0);
	}
	HashTable *table;
	FILE *fp;
	h = EVP_aes_128_ecb();
	ctx = EVP_CIPHER_CTX_new();
	char *rainbowFileName = argv[3];
	int s = atoi(argv[2]);
	l = atoi(argv[1]);
	rows = (16 * pow(2, s)) / (l*2);
	int k = ((pow(alphabetLen, l)/rows) + 1) * 2;

	printf("Output file: %s\n", rainbowFileName);
	printf("Num rows: %ld\n", rows);
	printf("Password length: %d\n", l);
	printf("s value: %d\n", s);
	printf("k value: %d\n", k);

	fp = fopen(rainbowFileName, "wb+");
	fprintf(fp, "%d%d\n", l, k);

	/**
	 * Rainbow table generation
	 */
	char pwd[l];
	char key[16];
	char out[16];
	unsigned __int128 universe = (unsigned __int128)pow(alphabetLen,l);
	srand(time(NULL));
	hashtable_t *ht = ht_create(rows);

	for(int i = 0; i < rows; i++)
	{
		//Generate Seed		
		while (1)
		{	
			int seed = rand() % (long)universe;
			getPwd(pwd, seed);
			if(!ht_get(ht, pwd))
				break;
		}
		fprintf(fp, "%s", pwd);
		ht_put(ht, pwd, pwd);

		//Compute
		for(int j = 0; j < k; j++)
		{
			//Key generation
			keyGen(key, pwd);

			//Cipher with AES
			int len;
			EVP_EncryptInit(ctx, h, key, 0);
			EVP_EncryptUpdate(ctx, out, &len, key, 16);
			EVP_CIPHER_CTX_cleanup(ctx);
		
			//Transform binary into a integer (long long)
			unsigned __int128 *hashValue = (unsigned __int128*)&out; 
			*hashValue += j;
			int newPwdValue = *hashValue % universe;
			getPwd(pwd, newPwdValue);
		}
		fprintf(fp, "%s", pwd);
	}
	fclose(fp);
}	

